Remaining work for restricted user access feature for guests.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First two commits are preparatory refactors.
- Next commit is to not send sender's original details in push notifications.
- Next commit is to not send sender's original details in message objects.
- Next commit is to allow restricted user access for guests only on plus plans.
- Next commit is to add client capability.
- Last two commits are to not send unknown users data to webapp and some UI changes.

![Screenshot from 2023-11-28 23-46-33](https://github.com/zulip/zulip/assets/35494118/9905ca56-d028-4139-b1a9-5a1db78e3888)
![Screenshot from 2023-11-28 23-45-24](https://github.com/zulip/zulip/assets/35494118/947290e6-c62e-49b0-8bc1-bbe0db029909)


<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/actions/message_send.py:
```diff
@@ -1122,6 +1124,19 @@ class UserData(TypedDict):
+
+            if user_access_restricted_in_realm(
+                send_request.message.sender
+            ) and not subscribed_to_stream(send_request.message.sender, send_request.stream.id):
```
We only allow sending a single at a time message using API and we only send multiple messages in development environment and probably for some notifications sent by bots, so this should not really do query in a loop. For bots the `user_access_restricted_in_realm` will anyways be `False` without doing any query.

timabbott replied:
OK -- let's add a comment though.

sahil839 commented on zerver/actions/realm_settings.py:
```diff
@@ -475,6 +476,17 @@ def do_change_realm_plan_type(
 
+    if old_value in [Realm.PLAN_TYPE_PLUS, Realm.PLAN_TYPE_SELF_HOSTED] and plan_type not in [
+        Realm.PLAN_TYPE_PLUS,
+        Realm.PLAN_TYPE_SELF_HOSTED,
```
We would not disable this feature in self-hosted realms, but I am not sure whether self-hosted orgs will use this code to downgrade the plan so we can probably remove that part.

sahil839 commented on zerver/lib/events.py:
```diff
@@ -132,6 +132,7 @@ def fetch_initial_state_data(
     spectator_requested_language: Optional[str] = None,
     pronouns_field_type_supported: bool = True,
     linkifier_url_template: bool = False,
+    supports_incomplete_realm_users: bool = False,
```
We are still discussing the name for this, so will add the documentation later.

sahil839 commented on web/src/people.ts:
```diff
@@ -1501,6 +1518,15 @@ export function extract_people_from_message(message: MessageWithBooleans): void
+            // This may lead to cases where an accessible user is
+            // not present in the users data for some time, but
+            // it will eventually be added later.
+            continue;
```
I am not sure about this approach, but we do not have any way to know whether the user ID not present in users data is invalid or is for a unknown user or just a accessible user for which the client does not have any data.

timabbott replied:
This is probably the wrong fix -- we need to call `_add_user` even without this feature just to avoid throwing unknown user exceptions.

Probably we don't need a change here at all, and instead we should just have `make_user` set a `inaccessible_user: bool` property that it's an "unknown user", and have the right sidebar logic just exclude such fake users in `filter_user_ids`, where we filter muted users and bot users.

sahil839 commented on zerver/lib/events.py:
```diff
@@ -397,6 +397,10 @@ def fetch_initial_state_data(
         state["server_supported_permission_settings"] = get_server_supported_permission_settings()
+
+        state[
+            "can_access_all_users_group_limits_presence"
```
Still need to add API documentation about this.

timabbott replied:
https://github.com/zulip/zulip/pull/27944/commits/6f613daabf7ffcfd1df525d0a7efaffd411e98da#r1408603397 probably means you don't need this in the API.

sahil839 commented on web/src/people.ts:
```diff
@@ -757,15 +758,24 @@ export function format_small_avatar_url(raw_url: string): string {
 export function sender_is_guest(message: Message): boolean {
     if (message.sender_id) {
-        const person = get_by_user_id(message.sender_id);
+        const ignore_missing = !settings_data.user_can_access_all_other_users();
```
We might also set `ignore_missing` as `True` for all cases and assume that the user ID would be valid. Anyway here as well we are atleast assuming that for guests and when the user access setting is set to "members".

timabbott replied:
OK -- let's add a comment to that effect that we'll find if grepping for the setting name, as that's a thing we'll certainly do if changing the permissions options.

sahil839 commented:
I guess we might need some changes in the popover. We can probably increase the height of name row.

alya commented:
Screenshots look good to me!

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -986,6 +1010,18 @@ def get_message_payload_gcm(
         user_profile, message, mentioned_user_group_id, mentioned_user_group_name
     )
+
+    if message.is_stream_message() and not check_can_access_user(message.sender, user_profile):
```
We should avoid doing this check in several places -- it's expensive, so doing 3-4 for a given push notifications API call can be expensive. It shouldn't be too hard to just pass a `can_access_sender` boolean down from an outer function.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -251,6 +255,15 @@ def accepts_event(self, event: Mapping[str, Any]) -> bool:
+            event["type"] == "realm_user"
+            and event["op"] == "add"
+            and self.supports_incomplete_realm_users
+            and event["person"]["full_name"] == str(UserProfile.INACCESSIBLE_USER_NAME)
```
Yeah this is not ideal; it might even just fail to work on account of i18n issues. One option is to put an extra field in the event that gets popped in a function along the lines of `process_custom_profile_fields_event` and used to know that this is an event that should only be put into Tornado queues that lack the client capability.

```
diff --git a/zerver/actions/create_user.py b/zerver/actions/create_user.py
index 51c4f54898..8f4b3a732f 100644
--- a/zerver/actions/create_user.py
+++ b/zerver/actions/create_user.py
@@ -394,6 +394,7 @@ def notify_created_user(user_profile: UserProfile, notify_user_ids: List[int]) -
             type="realm_user",
             op="add",
             person=get_data_for_inaccessible_user(user_profile.realm, user_profile.id),
+            inaccessible_user=True,
         )
         send_event_on_commit(user_profile.realm, event, user_ids_without_access_to_created_user)
 
```

timabbott commented on web/src/presence.ts:
```diff
@@ -206,7 +207,11 @@ export function set_info(
+                !(watchdog.suspects_user_is_offline() || reload_state.is_in_progress()) &&
+                (settings_data.user_can_access_all_other_users() ||
+                    page_params.can_access_all_users_group_limits_presence)
+            ) {
```
I have an alternative proposal here, which is to just remove this logging. I'm pretty sure we've fixed all the actual bugs this was good for finding, and it's basically spam now.

Can be a separate commit that just replaces this blueslip logging (i.e. just having a `continue` in this block) with a comment:

> There are a number of situations where it is expected that we get presence data for a user ID that we do not have in our user database, including when we're offline/reloading (watchdog.suspects_user_is_offline() || reload_state.is_in_progress()), when `CAN_ACCESS_ALL_USERS_GROUP_LIMITS_PRESENCE` is enabled, and whenever presence wins a race with the events system for events regarding a newly created or visible user.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -141,6 +143,7 @@ def to_dict(self) -> Dict[str, Any]:
         return dict(
             user_profile_id=self.user_profile_id,
             realm_id=self.realm_id,
+            realm_string_id=self.realm_string_id,
```
Can you split the change to add `realm_string_id` to a separate prep commit or maybe even PR? It involves a Tornado save/restore API change that require some migration care; see `TODO/compatibility` comments in this file for background/

Also, is that the right field to pass? `realm.host` has `REALM_HOSTS` transformations applied and so a `realm_host` field might be more correct.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1144,7 +1153,10 @@ def get_client_payload(apply_markdown: bool, client_gravatar: bool) -> Dict[str,
+        can_access_sender = client.user_profile_id not in user_ids_without_access_to_sender
+        message_dict = get_client_payload(
+            client.apply_markdown, client.client_gravatar, client.realm_string_id, can_access_sender
+        )
```
The negative is a good way to do it, to avoid sending extra copies of the whole user ID list.

timabbott commented:
This is progressing well. I haven't fulled reviewed the main commit for the `message` fetching stuff, which seems the most complex, but I think most of the other commits look good with the tweaks suggested above.

sahil839 commented on web/src/activity_ui.js:
```diff
@@ -137,6 +137,15 @@ export function initialize(opts) {
+    // disabled. We just ignore that event and return.
+    const person = people.maybe_get_user_by_id(user_id, true);
+    if (person === undefined || person.is_inaccessible_user) {
+        return;
```
I have not added checks here like `settings_data.user_can_access_all_other_users` or check for `CAN_ACCESS_ALL_USERS_GROUP_LIMITS_PRESENCE` assuming that this would only happen for an inaccessible user.

timabbott replied:
Can you talk about what motivates a change here at all (i.e. what happens if we fall through?)

alya commented:
Something isn't working right with @-mentions. Here, Desdemona is a former subscriber to this stream, whereas Polonius and Zoe were never subscribed. Clicking on Polonius and Zoe produces errors:

![Screenshot 2023-12-03 at 11 21 27 PM](https://github.com/zulip/zulip/assets/2090066/f7ab6c9b-e69d-4ffe-8dfb-94e4c0e73cdb)

alya commented:
I guess the setting still needs to be renamed.

sahil839 commented:
Updated the PR with UI changes. I have some behavioral questions here -
- We do not allow muting/unmuting inaccessible users in API. How do we want to handle "Unmuting" part in frontend? We use a link like text for unmuting in the "Muted users" list.
- For now, I have added the commit for guests to not see typing notifications from inaccessible users. Are we fine with this or we want to show that as "Unknown user is typing"? I don't think showing it will be helpful.
- How do we want to handle restoring drafts and editing scheduled messages to inaccessible users? For both cases we add the recipient user pills in compose box. I guess we can just disallow the action. There is already a bug when trying to restore a draft for inaccessible stream.
- I am not sure about how this popover looks without user circle for inaccessible users.
![image](https://github.com/zulip/zulip/assets/35494118/4c7317fe-5200-4f4c-a55f-57f777a5e055)

sahil839 commented on web/src/people.ts:
```diff
@@ -1470,6 +1470,17 @@ export function make_user(user_id: number, email: string, full_name: string): Us
+
+    const email = "user" + user_id + "@" + page_params.realm_bot_domain;
+    const unknown_user = make_user(user_id, email, INACCESSIBLE_USER_NAME);
+    _add_user(unknown_user);
```
Do we want to add this user to `active_user_dict` as well? We do have `is_active` as `true` for all inaccessible users. Asking this as we do not do it when making fake user objects for message senders.

timabbott replied:
I think it would basically be reasonable to just make a fake user using `add_inaccessible_user` whenever we are processing a `user_id` that the server assures us must exist (i.e. something the server gave us, not like a value put into a potentially bogus narrow URL). So I'm not sure why we wouldn't want to do this in the message sender code path and emoji author code paths too.

timabbott replied:
I think it might make sense to abstract this more nicely with just having a `get_user_by_id_assert_valid()` function that calls `add_accessible_user` if the user doesn't exist -- a lot of these little commits, like the bots one, where we wouldn't necessarily need special logic, might then be just replacing `get_user_by_id` with `get_user_by_id_assert_valid` without broader changes.

sahil839 replied:
Ok, I got this. Will update the PR to create the fake user objects for all `user_id` provided by server.

But the question that I was asking above was that before this PR as well, we created fake user objects for the users who are sender of the messages received by client but the client has not received the data for them. In that case we do not add the user in `active_user_dict`, which we use at certain places to determine if the user is active or not.

So, do we want to change this behavior now to add the users in `active_user_dict` as well?

timabbott replied:
Yeah I think that's the right plan.

Next year we'll do some API design discussions about reworking how "display recipients" work, which may involve changing the overall policy here, but I don't see any technical problem with the current approach.

sahil839 commented on web/src/rendered_markdown.js:
```diff
@@ -139,11 +139,22 @@ export const update_elements = ($content) => {
+                // the "Unknown user" popover is opened on clicking
+                // the mention pill.
+                if (person === undefined) {
+                    people.add_inaccessible_user(user_id);
```
I am not sure how to handle the case when the markdown element only has `data-user-email` and not `data-user-id`. I don't think we would want to create fake user objects with no or some random ID. Or do we just add separate code to show the "Unknown user" in the click handler, without creating any user object.

timabbott replied:
Such cases are super rare and very old (1137ef4e40258280e202325c9295579c93fb948b). All we need is for the client to not crash on displaying them -- It's probably fine if clicking on such an ancient mention doesn't open the popover, for example -- that will realistically probably never happen again for those 6+ year old mentions.

sahil839 commented:
Regarding creating user objects for inaccessible users, we currently only do so if we expect the user to be shown as a pill or a popover. And for rest of the cases, we do not create them.

alya commented:
> For now, I have added the commit for guests to not see typing notifications from inaccessible users. Are we fine with this or we want to show that as "Unknown user is typing"? I don't think showing it will be helpful.

This seems fine to me. I don't even know how it would come up.

alya commented:
> I am not sure about how this popover looks without user circle for inaccessible users.

That alignment looks too far to the left. I'd left-align "Unknown user" with the circle.

alya commented:
> We do not allow muting/unmuting inaccessible users in API. How do we want to handle "Unmuting" part in frontend? We use a link like text for unmuting in the "Muted users" list.

How about removing the "Unmute" link in that table?

alya commented:
I left a question [here](https://chat.zulip.org/#narrow/stream/101-design/topic/access.20to.20all.20users.20in.20realm.20.2310970/near/1696180).

sahil839 commented:
> > How do we want to handle restoring drafts and editing scheduled messages to inaccessible users? For both cases we add the recipient user pills in compose box. I guess we can just disallow the action. There is already a bug when trying to restore a draft for inaccessible stream.
> 
> Can we remove the recipients when you restore the draft, or something along those lines?

I checked just now and we handle the restoring the draft case like that only. Do we want to do the same for scheduled messages as well?

sahil839 commented:
> > We do not allow muting/unmuting inaccessible users in API. How do we want to handle "Unmuting" part in frontend? We use a link like text for unmuting in the "Muted users" list.
> 
> How about removing the "Unmute" link in that table?

The user would still be allowed to unmute some of the users. It would look something like this -
![image](https://github.com/zulip/zulip/assets/35494118/223c8fcb-18e4-46f1-9b44-37d12e1c22dd)

alya commented:
An "Unknown user" pill is basically fine too; I don't think it super matters either way. But it's weird for it to say that the user was deactivated when they hadn't been...

alya commented:
> The user would still be allowed to unmute some of the users. It would look something like this -

Yeah, that seems fine to me.

timabbott commented on zerver/lib/mention.py:
```diff
@@ -105,9 +108,15 @@ def get_full_name_info_list(self, user_filters: List[UserFilter]) -> List[FullNa
+            possible_mention_user_ids = [row.id for row in rows]
+            inaccessible_user_ids = get_inaccessible_user_ids(
+                possible_mention_user_ids, message_sender
+            )
```
I this doesn't work correctly if the user they are trying to mention is in the `user_filters` cache.

sahil839 replied:
Ohh, forgot to comment on this.

For sending messages through API and most of the other cases, we always create a new `MentionBackend` object for each message so `user_cache` is initialized for each message. This happens for most of the cases except when we send bot notifications for stream and group subscriptions.

timabbott replied:
I see, is another framing that the cache is never reused with multiple senders?

timabbott commented on web/src/settings_data.ts:
```diff
@@ -246,6 +246,17 @@ export function bot_type_id_to_string(type_id: number): string | undefined {
+export function user_can_access_all_other_users(): boolean {
+    if (!page_params.user_id) {
+        return true;
+    }
```
This is the spectators case, right?

sahil839 commented:
Ohh, so it is not working correctly when there is a user object for the inaccessible user.

I am aware of the "deactivated" bug. Will see how easy it to implement the no user pill fix and will update accordingly.

timabbott commented on web/src/settings_emoji.ts:
```diff
@@ -111,10 +111,16 @@ export function populate_emoji(): void {
+                emoji.author = {full_name: people.INACCESSIBLE_USER_NAME};
+            } else {
+                emoji.author = {full_name: author.full_name};
+            }
```
This might break live-update if an author's name changes for a real user. 

I feel like we could just generate a fake user object using `make_user` to point to instead of having custom logic in this code path?

timabbott commented:
I pulled out the first few commits including the API change (plus some reordered cleanup commit) to https://github.com/zulip/zulip/pull/28062, and marked that to merge once CI passes on it.

In terms of the state of the review:
* https://github.com/zulip/zulip/pull/27944#discussion_r1416750835 is probably most significant to think through, as it might reduce duplicated code and complexity for the 50% or so of the later commits where we're not doing much particularly special for inaccessible users. (Obviously things like the groups typeahead stuff will still be important).
* I skipped reading the `message` commit due to a shortage of time. We probably need to do something like that, but it's relatively performance-sensitive, and I'll want to study it clearly. I think it's more or less the last performance-sensitive backend change, so it could make sense to just pull it out to its own PR.

timabbott commented:
> I am not sure about how this popover looks without user circle for inaccessible users.

Could we just treat inaccessible users as always offline for display in UI like that? It seems much more robust than doing custom styling just for them.

alya commented:
Sure, that seems fine.

alya commented:
> There is one more bug where we allow users to add inaccessible users in user type custom profile field by manually typing the email. Will fix this soon.

I think it's fine for this sort of obscure case to be a follow-up, even if we have to fix it in 8.1.

sahil839 commented on web/src/user_group_popover.js:
```diff
@@ -117,14 +117,19 @@ export function register_click_handlers() {
+            // We need to include inaccessible users here separately, since
+            // we do not include them in active_user_dict, but we want to
+            // show them in the popover as "Unknown user".
+            .filter((m) => people.is_active_user_for_popover(m.user_id) || m.is_inaccessible_user)
```
Added the `m.is_inaccessible_user` condition since we do not include inaccessible users in `active_user_dict`.

sahil839 replied:
Also done similar change in the last commit to not show "deactivated" in user pills.

sahil839 commented on web/src/people.ts:
```diff
@@ -1242,9 +1252,19 @@ export function filter_people_by_search_terms(
+        if we would want to use this function for other
+        cases where we might want to display inaccessible
+        users as "Unknown user".
+    */
```
We also not show the mention pill if user types `@**Unknown user|{user_id}**` when `user_id` is ID of an inaccessible user.

sahil839 commented:
Updated the PR -
- We now do not show "deactivated" in pills for inaccessible users.
- We show presence circle for inaccessible users in group popover.
- Added commit to not render mention pill for inaccessible users while local echoing messages and in drafts as well.

Regarding -
> There is one more bug where we allow users to add inaccessible users in user type custom profile field by manually typing the email. Will fix this soon.

This bug is also present in compose box DM recipient. This will need some changes in the pills code and is not straightforward so will fix this as a follow-up.

timabbott commented on web/src/composebox_typeahead.js:
```diff
@@ -516,8 +518,19 @@ export function get_person_suggestions(query, opts) {
+        groups = user_groups.get_realm_user_groups().filter((group) => {
+            const group_members = group.members;
+            for (const user_id of group_members) {
+                const person = people.maybe_get_user_by_id(user_id, true);
```
That seems OK.

timabbott commented on static/images/unknown-user-avatar-medium.png:
```diff
@@ -0,0 +1 @@
+unknown-user-avatar.png
```
Is it easy to add a test that accessing the `-medium.png` URL works?

sahil839 replied:
I cannot think of any way to test static file URLs.

timabbott commented:
I extracted two of the commits to https://github.com/zulip/zulip/pull/28113 and marked to merge once CI passes, after adding a comment to explain why the mention commit is correct.

timabbott commented:
Added a couple comments to the last commit and merged, thanks @sahil839!
